### ISSUE TITLE
Insert #include <winrt/Microsoft.UI.Xaml.Controls.h> into pch.h not only in Main.h

### ISSUE THREAD

**Issue by 0x000eHx** [MicrosoftDocs/windows-uwp#2280]
Hey there!
tl;dr:
I would suggest to correct this article by the instructions of the linked article from your Microsoft Documentation:
https://docs.microsoft.com/en-us/uwp/toolkits/winui/getting-started

This article stats to put the #include &lt;winrt/Microsoft.UI.Xaml.Controls.h&gt; (and so on) into the Main.h file.
The linked article says to put them into the pch.h which is as I assume the better choice.

long story:
I followed the explanations of this article and got some weird errors with Objects of the type `::winrt::Microsoft::UI::Xaml::Controls::Numberbox`.
Error message:
`'Microsoft': the symbol to the left of a '::' must be a type`
so the Numberbox specifically is kinda not present.
Using the Control Element Numberbox is essential to my project, so I wondered why this Element was "not known" even when I installed the latest preview Version of the NuGet Package Microsoft.UI.Xaml and inserted everything into the App.xaml, MainPage.xaml and Main.h in a freshly made C++/WinRT Project (ofc WinRT VSIX is installed).
The Control Numberbox is in the WinUI Library since version 2.3, so the preview Version of 2.4 should be fine too...
Inserting the #include &lt;winrt/Microsoft.UI.Xaml.Controls.h&gt; into the pch.h instead of Main.h earlier in the generating process solved these issues.

Thanks for the effort.
Greetings

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: eb0978bf-4d67-556c-b6c6-87b360d446c8
* Version Independent ID: 6fe15723-a3a2-9a7c-1720-ef4194f5ed70
* Content: [A simple C++/WinRT Windows UI Library example - UWP applications](https://docs.microsoft.com/en-us/windows/uwp/cpp-and-winrt-apis/simple-winui-example)
* Content Source: [windows-apps-src/cpp-and-winrt-apis/simple-winui-example.md](https://github.com/MicrosoftDocs/windows-uwp/blob/docs/windows-apps-src/cpp-and-winrt-apis/simple-winui-example.md)
* Product: **uwp**
* Technology: **platform**
* GitHub Login: @stevewhims
* Microsoft Alias: **stwhi**

**Comment by stevewhims** [comment]
Hi, @0x000eHx, and thanks for filing this issue. I have just now published updates and clarifications to the topic, so those are live now. I'll outline here the changes I made, and the changes I didn't make, and I'll talk about my reasons for each.

First, the topic used to say (about the includes in `MainPage.h`) "they cause those projection header files to be included in your project". I would say that they're not included *in your project*, but rather in **MainPage**. I thought that was misleading enough that I wanted to correct my wording there.

I've also added an "Important" callout, hopefully focusing on the difference between including at the file level and including at the precompiled header level. I do feel like for this particular walkthrough, including locally like this is appropriate. It's a valid choice in any real-world project, and I'd say it's the better choice in *some*. For example, imagine you have a large project, with a set of very commonly used headers included in `pch.h`. You make changes to `pch.h` very rarely at this point in your project's lifecycle and so each time you build you benefit from not having to recompile the headers in your `pch.h`. Now imagine that you decide to use just one a WinUI type in just one of your XAML pages. If you add the headers to `pch.h` and build then you'll have to wait for `pch.h` to re-build, which will take a good chunk of time. If you add the headers just to the local XAML file, then your build will (typically) be a lot faster.

So, including locally or globally are both valid choices. They will have different values and tradeoffs for different situations. The other topic that you link to (the C++/WinRT content in both these topics was written by the same author) shows one option, and doesn't discuss the other (which is fine for the level it's pitched at). This topic now explains the difference better, so thanks for the prod there.

And just to double-check, I went through the walkthrough in this topic another time, and then I added an instance of a **NumberBox** (note **NumberBox** and not **Numberbox**) declaratively to the XAML markup, and then imperatively to the `.h` and `.cpp`, and the references to that type resolved in all cases.

Thanks again!
Steve

**Comment by 0x000eHx** [comment]
Thanks for the fast reply and clarification 👍 
You are totally right, I did recall that namespace path of Number**b**ox out of memory, after I sorted out that issue. It must be Number**B**ox.

I'm curious why my simple project does produce this error. I only use the NumberBox Element inside my MainPage. Means only inserted it inside the MainPage.xaml in a one page MainPage Project.
Dont really know why the includes in the MainPage.h and .xaml weren't suffient...
Did you use any other #includes than `#include "winrt/Microsoft.UI.Xaml.Controls.h"` and `#include "winrt/Microsoft.UI.Xaml.XamlTypeInfo.h"` in the MainPage.h File? Is the order of insertation important? (Did insert these entries after the `#include "MainPage.g.h"` tho')
What was your assigned Windows SDK / Target Platform Version? My settings are the standard ones: target 10.0.18362.0 and minimum version 10.0.17763.0

By the way I had to include the NumberBox XAML Code manually into the MainPage.xaml, cause it won't show as an available element in the Toolbox of Blend... Was it the same in your example build?
This might not be on purpose and I have to direct this as an "error" to the WinUI NuGet Package Devs? Maybe this will change in future updates...

Thanks again for your time.
Greetings

**Comment by stevewhims** [comment]
I only used the #includes mentioned in the topic, order of include doesn't matter AFAIK, but I used the order in the topic. For all sample code I put into topics it's always tested with whatever the currently generally available platform is, for both target and min (currently that's 10.0.18362.0). I'm unsure how Visual Studio/Blend behave in this case.

-Steve
